### PR TITLE
[Test] Eager cancelation for ProgramTester

### PR DESCRIPTION
**Pull request by EronWright**
# Description

This PR changes the behavior of `ProgramTester` in `pkg/testing/integration` package, to detect test failures more eagerly and then to skip remaining steps. Specifically it checks whether the test been marked as failed by a validation function (as defined by [`t.Failed()`](https://pkg.go.dev/testing#T.Failed)).

The rationale is that, for most integration tests, each step assumes that the previous step was successful. It is simply noisy to continue with the subsequent steps. A workaround is to call `t.FailNow()` in the validation function (or use `require`).

An option is provided `ExpectTestFailure` to continue with steps after test failure (as is the current behavior).

There exists an option `ExpectFailures` to continue with steps after a pulumi deployment failure.  I took such failures to be orthogonal to test failure.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version
  

## Example
Here's a snippet of output in the event that a validation function marks the test as failed (e.g. using an assertion).
```
    ...
    /Users/eronwright/Pulumi/pulumi/tests/examples/program.go:1742: Performing extra runtime validation.
    /Users/eronwright/Pulumi/pulumi/tests/examples/examples_test.go:44:
                Error Trace:    ...
                Error:          Expected nil, but got ...
                Test:           TestAccMinimal
    /Users/eronwright/Pulumi/pulumi/tests/examples/program.go:1744: Extra runtime validation complete.
    /Users/eronwright/Pulumi/pulumi/tests/examples/program.go:1164: Canceling further steps due to test failure
    /Users/eronwright/Pulumi/pulumi/tests/examples/program.go:1349: Destroying stack
    ...
    /Users/eronwright/Pulumi/pulumi/tests/examples/program.go:1365: Test failed, retaining stack '...'
--- FAIL: TestAccMinimal (5.61s)
```

### CODE REVIEW

File: pkg/testing/integration/program.go [hunk 1]
```diff
@@ -68,6 +68,8 @@ const (
 
 const windowsOS = "windows"
 
+var ErrTestCanceled = errors.New("test canceled")
```
Comment (tgummerer) [inline pkg/testing/integration/program.go:71]: I'm bikeshedding here, but maybe `ErrTestAlreadyFailed` would be a better name?  canceled sounds a bit like something a user did to me.

**Review by tgummerer — COMMENTED** [review]
> The rationale is that, for most integration tests, each step assumes that the previous step was successful. It is simply noisy to continue with the subsequent steps. A workaround is to call `t.FailNow()` in the validation function.

Out of curiosity, what's the rationale for not just calling `FailNow()` and having this mechanism instead?  If we call `FailNow()` at the point where we no longer care about any additional failures we could potentially save a developer some time by not running the rest of the steps instead of running them anyway and just not reporting the errors.  

Is there some advantage to running the rest of the steps without reporting the errors?

**Comment by Frassle** [comment]
> Out of curiosity, what's the rationale for not just calling FailNow() and having this mechanism instead?

We've been pushing to move a lot of our test checks to use "github.com/stretchr/testify/require" instead of "github.com/stretchr/testify/assert" for failures that would invalidate the rest of the test. 

I think I'd prefer that approach here rather than adding yet another option to ProgramTest.

File: pkg/testing/integration/program.go (hunk 1 above)
Comment (EronWright) [reply pkg/testing/integration/program.go:71]: I considered something similar (`ErrTestFailed`) but I convinced myself that the purpose of the error is to interrupt the flow of the `ProgramTest` func without logging the error.  Still I think I'll go back to `ErrTestFailed`.

**Comment by Frassle** [comment]
> if you would prefer we could eliminate the ExpectTestFailure option, since I'm not aware of any actual use case for it. WDYT?

Yeh I'd be ok with that.

File: pkg/testing/integration/program.go [hunk 2]
```diff
@@ -792,7 +794,9 @@ func ProgramTest(t *testing.T, opts *ProgramTestOptions) {
 	prepareProgram(t, opts)
 	pt := newProgramTester(t, opts)
 	err := pt.TestLifeCycleInitAndDestroy()
-	assert.NoError(t, err)
+	if !errors.Is(err, ErrTestFailed) {
```
Comment (Frassle) [inline pkg/testing/integration/program.go:797]: Probably worth a quick comment here that this is so we don't trigger an assert message when we've already failed an assert.

**Review by Frassle — APPROVED** [review]

**Pull request merged**